modules/library/metadata/src/main/java/org/geotools/util/SoftValueHashMap.java

```diff
@@ -562,13 +511,12 @@ public String toString() {
          */
         @SuppressWarnings("unchecked")
         private boolean findNext() {
-            assert Thread.holdsLock(hash);
             while (iterator.hasNext()) {
                 final Map.Entry<K,Object> candidate = iterator.next();
                 Object value = candidate.getValue();
                 if (value instanceof Reference) {
-                    value = ((Reference) value).get();
-                    entry = new MapEntry<K,V>(candidate.getKey(), (V) value);
+                    value = ((Reference<K, V>) value).get();
+                    entry = new MapEntry<K,V>((K) SoftValueHashMap.replaceNull(candidate.getKey()), (V) value);
```

Comment:
Shouldn't this be the other way around? if the placeholder is used, replace it with null instead? Is there a test to cover this, if not, please add one.